pytest_mergify/ci_insights.py

```diff
@@ -220,46 +233,124 @@ def _fetch_existing_test_names(self) -> typing.List[str]:
+                )
+
+                message += (
+                    f"{os.linesep}- '{test_name}' has been tested {retry_count} "
+                    f"times using approx. {test_retry_duration_ms / budget_duration_ms * 100:.1f} % "
+                    f"of the budget ({test_retry_duration_ms} ms/{budget_duration_ms} ms)"
+                )
+        else:
+            message += f"{os.linesep}- No new tests detected, but we are watching 👀"
+
+        terminalreporter.write_line(message)
+
     def handle_flaky_detection_for_report(
         self,
         report: _pytest.reports.TestReport,
     ) -> None:
-        if not self.is_flaky_detection_active():
+        if not self._is_flaky_detection_active():
             return
 
         if report.outcome not in ["failed", "passed"]:
             return
 
         test_duration_ms = int(report.duration * 1000)
-        self.total_test_durations_ms += test_duration_ms
+        self._total_test_durations_ms += test_duration_ms
 
         test_name = report.nodeid
-        if test_name in self.existing_test_names:
+        if test_name in self._existing_test_names:
             return
 
-        self.add_new_test_duration(test_name, test_duration_ms)
+        self._add_new_test_duration(test_name, test_duration_ms)
 
         if self.tracer:
             opentelemetry.trace.get_current_span().set_attributes(
                 {"cicd.test.new": True}
             )
 
     def run_flaky_detection(self, session: _pytest.main.Session) -> None:
-        if not self.is_flaky_detection_active():
+        if not self._is_flaky_detection_active():
             return
 
-        for item in session.items:
-            if item.nodeid not in self.new_test_durations_by_name:
-                continue  # Only run for newly added tests.
+        new_items = [
+            item
+            for item in session.items
+            if item.nodeid in self._new_test_durations_by_name
+        ]
 
-            self._run_flaky_detection_for_item(item)
+        budget_deadline = datetime.datetime.now(
+            datetime.timezone.utc
+        ) + datetime.timedelta(milliseconds=self._get_budget_duration_ms())
 
-    def _run_flaky_detection_for_item(
-        self,
-        item: _pytest.nodes.Item,
-    ) -> None:
-        for _ in range(5):  # NOTE(remyduthu): Hard-coded value for now.
-            item.ihook.pytest_runtest_protocol(item=item, nextitem=None)
+        while datetime.datetime.now(datetime.timezone.utc) <= budget_deadline:
+            tests_to_retry = [
+                item for item in new_items if self._should_retry_test(item.nodeid)
+            ]
+            if not tests_to_retry:
+                break
+
+            # NOTE:
+            #   We are running each test once at each iteration until the budget
+            #   is exhausted.
+            for item in new_items:
+                if not self._should_retry_test(item.nodeid):
+                    continue
```

Comment:
maybe i am missing something, but i think we can use the `tests_to_retry` defined previously 🤔 


```suggestion
            for item in tests_to_retry:
```

Reply:
Oh wow, yes, thanks!